RFCs/rfc-0009-wallet-provider-store-request-api.md

```diff
@@ -0,0 +1,247 @@
+
+#### Primary Functionalities:
+The proposed API should empower a Mina wallet to:
+- **Receive Objects**: Accept and store objects, whether they're verifiable credentials, proofs, or any other data.
+- **Request Objects**: Initiate requests to retrieve stored objects, facilitating interactions like transaction validations or data computations.
+
+### Scenario: A Wallet Receiving Objects to Store
+#### `mina_setState`
+Providers within Mina's domain could introduce a new RPC method: `mina_setState`. When called, a user interface might emerge, allowing users to choose whether to store objects from a zkApp. This method returns a Promise: resolving with a boolean or, if storing is unavailable, rejecting with an Error. A zkApp MUST provide sufficient information in the request to allow the wallet to store and index the object appropriately.
+
+zkApps may need to store objects in a wallet, such as verifiable credentials or proofs, which are critical for subsequent interactions. To facilitate this, the wallet requires a new provider method with specific parameters:
+
+```
+// Pseudocode
+START zkApp
+IF provider is defined
+    REQUEST[1] object storage
+    IF user approves
+        RESOLVE[2] object stored
+        CONTINUE zkPP
+    IF user rejects
+        REJECT[3] object storage
+        STOP zkApp
+IF provider is undefined
+    STOP zkApp
+```
+[1] `REQUEST`
+zkApp MUST request object storage by calling the `mina_setState` RPC method on the provider exposed at `window.mina`. Calling this method MUST emit an event `message` that signifies the zkApp's initiation of the process to request the wallet to store an object and MAY trigger a user interface that allows the user to approve or reject object storage for a given zkApp. This method MUST return a Promise that is resolved with a boolean or rejected if no storage is available (e.g., the user rejected object storage).
+
+ [2] `RESOLVE`
+The Promise returned when calling the `mina_setState` RPC method MUST be resolved with a boolean. This must emit an event `message` indicating that the request to store the object has been approved and the promise has been resolved with a boolean.
+
+ [3] `REJECT`
+The Promise returned when calling the `mina_setState` RPC method MUST be rejected with an informative Error if no storage is available for any reason. This must emit an event `message` indicating that the request to store the object has been rejected, and the promise is rejected with an error.
+
+A wallet MUST adequately detail to the user what the zkApp is requesting the wallet to store. If approved the wallet MUST appropriately store the object for future use. The zkApp COULD provide in the params appropriate information for how this object is indexed if the zkApp will require the information later.
+
+#### Example initialization - Request objects to be stored in the wallet
```

Comment:
Is it possible this is not a pattern we would want to allow? It seems if the information is private, we might not want to release the information to the website at all (vs releasing proofs to the website ie the attestation standard API)

Or are there cases where there is private data, that we do want websites to be able to request directly, and can't / don't want to just share a proof?

Reply:
This pattern would be for public, not sensitive information. Instead of a KYC credential that might contain personally identifiable information, a wallet might store a zkOracle attestation that makes a claim about some Web 2 data, in that case the user doesn't need to attest to that data again, it can just provide the original attestation proof it knows of.

In any case, a wallet should communicate to the user what a zkApp is requesting to ensure the user is completely aware of what data they are sharing. Data stored in a wallet should be encrypted at rest to prevent unauthorised access to it, and partitioned well enough to separate concerns and not accidentally provide private data.

Reply:
From my understanding the data could be stored as:

app_id  - data 

potentially on a merkle tree with app_id as index, 
this could enable even easier data provision into an app that waits for a merkle witness and data

Reply:
How does zkApp know whether the wallet has the information he needs? And how does the wallet know what information zkApp needs?

Reply:
> This pattern would be for public, not sensitive information. Instead of a KYC credential that might contain personally identifiable information, a wallet might store a zkOracle attestation that makes a claim about some Web 2 data, in that case the user doesn't need to attest to that data again, it can just provide the original attestation proof it knows of.
> 
> In any case, a wallet should communicate to the user what a zkApp is requesting to ensure the user is completely aware of what data they are sharing. Data stored in a wallet should be encrypted at rest to prevent unauthorised access to it, and partitioned well enough to separate concerns and not accidentally provide private data.

What would "the original attestation proof it knows of" look like? Is this caching proofs from the attestation API? Or something more general? Think I"m missing what other kinds of things this would want to expose.